streaming/src/main/scala/org/apache/spark/streaming/dstream/SocketInputDStream.scala

```diff
@@ -52,46 +52,60 @@ class SocketReceiver[T: ClassTag](
     storageLevel: StorageLevel
   ) extends Receiver[T](storageLevel) with Logging {
 
+  private var socket: Socket = _
+
   def onStart() {
-    // Start the thread that receives data over a connection
-    new Thread("Socket Receiver") {
-      setDaemon(true)
-      override def run() { receive() }
-    }.start()
+
+    try {
+      logInfo(s"Connecting to $host:$port")
+      socket = new Socket(host, port)
+      logInfo(s"Connected to $host:$port")
+
+      // Start the thread that receives data over a connection
+      new Thread("Socket Receiver") {
+        setDaemon(true)
+        override def run() { receive() }
+      }.start()
+    } catch {
+      case e: ConnectException =>
+        restart(s"Error connecting to $host:$port", e)
+      case NonFatal(e) =>
+        logWarning("Error receiving data", e)
+        restart("Error receiving data", e)
+    } finally {
+      onStop()
+    }
   }
 
   def onStop() {
-    // There is nothing much to do as the thread calling receive()
-    // is designed to stop by itself isStopped() returns false
+    //in case restart thread close it twice
+    if (socket != null) {
+      socket.close()
```

Comment:
This is now incorrect, since you can have two threads enter this block and both close the socket, or else close and null it before the other does. The synchronization of close() is irrelevant.